eventsourcing/domain/model/events.py

```diff
@@ -275,13 +275,21 @@ def unsubscribe(handler, predicate=None):
 
 def publish(event):
     matching_handlers = []
-    for predicate, handlers in _event_handlers.items():
-        if predicate is None or predicate(event):
-            for handler in handlers:
-                if handler not in matching_handlers:
-                    matching_handlers.append(handler)
-    for handler in matching_handlers:
-        handler(event)
+    events = []
+    if isinstance(event, (list, tuple)):
+        events = event
+    else:
+        events.append(event)
+
+    for event in events:
+        for predicate, handlers in _event_handlers.items():
+            if predicate is None or predicate(event):
+                for handler in handlers:
+                    if handler not in matching_handlers:
+                        matching_handlers.append(handler)
+        for handler in matching_handlers:
+            handler(event)
+
```

Comment:
Yes, but so that all the pending events in an AggregateRoot can be persisted in the same database transaction, we would still require the list of events to be passed to the handlers as a list.

One idea would be to change the ``publish()`` function, so that if it gets a list of events, then it calls each predicate to filter events, finally passing a filtered list to the handler. Something like this:

```
def publish(event):
    for predicate, handlers in _event_handlers.items():
        if isinstance(event, (list, tuple)):
            if predicate is None:
                filtered = event[:]
            else:
                filtered = [e for e in event if predicate(e)]
            if filtered:
                for handler in handlers:
                    handler(filtered)
        else:
            if predicate(event):
                for handler in handlers:
                    handler(event)
```

But that seems to make the publish function partly responsible for deciding which handlers are called: and predicates can only make decisions about individual events, which seems restrictive. On balance, I reckon it's just better for the predicate to get a list, and for the handlers also to get a list so they can do the filtering if they want to. I guess the predicate could be changed to return a list of booleans, one for each of the events in the list, so that only those events would be published to the companion subscriber, but it seems complicated. If the predicate and the handler both get the published list, both predicates and handlers get exactly what is published, and they can do whatever they need to do.

On reflection, the matching_handlers stuff in the ``publish()`` function doesn't really contribute anything. The ``publish()`` function could just be like this:

```
def publish(event):
    for predicate, handlers in _event_handlers.items():
        for handler in handlers:
            if predicate is None or predicate(event):
                handler(event)
```

Then, to get the behaviour I think you expected from the ``subscribe_to`` decorator, it could be changed to filter the published list a handler defined and subscribed by the decorator, without a predicate, which gets everything that is published, and calls the decorated function once for each event that matches the given class:

```
def subscribe_to(event_class):
    def wrap(func):
        def handler(event):
            if isinstance(event, (list, tuple)):
                for e in event:
                    if isinstance(e, event_class):
                        func(e)
            elif isinstance(event, event_class):
                func(event)
        subscribe(handler=handler)
        return func
    return wrap
```

I just changed the tests to cover this, and everything seems to work:

```
    def test_subscribe_to_decorator(self):
        entity_id1 = uuid4()
        event1 = Example.Created(
            originator_id=entity_id1,
            originator_topic=get_topic(Example),
            a=1, b=2
        )
        event2 = Example.Discarded(
            originator_id=entity_id1,
            originator_version=1,
        )
        handler = mock.Mock()

        # Check we can assert there are no event handlers subscribed.
        assert_event_handlers_empty()

        @subscribe_to(Example.Created)
        def test_handler(e):
            """Doc string"""
            handler(e)

        # Check the decorator doesn't mess with the function doc string.
        self.assertEqual('Doc string', test_handler.__doc__)

        # Check can fail to assert event handlers empty.
        self.assertRaises(EventHandlersNotEmptyError, assert_event_handlers_empty)

        # Check event is received when published individually.
        publish(event1)
        handler.assert_called_once_with(event1)

        # Check event of wrong type is not received.
        handler.reset_mock()
        publish(event2)
        self.assertFalse(handler.call_count)

        # Check a list of events can be filtered.
        handler.reset_mock()
        publish([event1, event2])
        handler.assert_called_once_with(event1)

        handler.reset_mock()
        publish([event1, event1])
        self.assertEqual(2, handler.call_count)

        handler.reset_mock()
        publish([event2, event2])
        self.assertEqual(0, handler.call_count)

```

If you wanted to submit these changes, I would definitely merge them in ;-).

Reply:
An alternative implementation of ``@subscribe_to`` would be to call the decorated function once with a filtered list of matching items, rather than once for each matching item in the list. At least then the list wouldn't always be disintegrated. Perhaps disintegrating the list could be an optional feature of the decorator, which by default calls the decorated function once with a filtered list?

Reply:
Oh ok now i see, `PersistencePolicy` also is using subscription to trigger save.